lib/charms/tls_certificates_interface/v0/tls_certificates.py

```diff
@@ -0,0 +1,417 @@
+        relation_data[certificate["common_name"]] = json.dumps(certificate_dict)
+
+        certificate_dict = {"key": certificate["key"], "cert": certificate["cert"]}
+        relation_data[certificate["common_name"]] = json.dumps(certificate_dict)
+
+    def _on_relation_changed(self, event):
+        relation_data = _load_relation_data(event.relation.data[event.unit])
+        if not relation_data:
+            logger.info("No relation data - Deferring")
+            event.defer()
+            return
+        if not self._relation_data_is_valid(relation_data):
+            logger.info("Relation data did not pass JSON Schema validation - Deferring")
+            event.defer()
+            return
+        for server_cert_request in relation_data.get("cert_requests", {}):
+            self.on.certificate_request.emit(
+                common_name=server_cert_request.get("common_name"),
+                sans=server_cert_request.get("sans"),
+                cert_type="server",
+                relation_id=event.relation.id,
+            )
+        for client_cert_requests in relation_data.get("client_cert_requests", {}):
+            self.on.certificate_request.emit(
+                common_name=client_cert_requests.get("common_name"),
+                sans=client_cert_requests.get("sans"),
+                cert_type="client",
+                relation_id=event.relation.id,
+            )
+
+
+class TLSCertificatesRequires(Object):
+    """TLS certificates requirer class to be instantiated by TLS certificates requirers."""
+
+    on = CertificatesRequirerCharmEvents()
+
+    def __init__(
+        self,
+        charm,
+        relationship_name: str,
+        common_name: str = None,
+        sans: list = None,
+    ):
+        super().__init__(charm, relationship_name)
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+        self.relationship_name = relationship_name
+        self.charm = charm
+        self.common_name = common_name
+        self.sans = sans
+
+    def request_certificate(
+        self,
+        cert_type: Literal["client", "server"],
+        common_name: str,
+        sans: list = None,
+    ):
+        """Request TLS certificate to provider charm.
+
+        :param cert_type: Certificate type: "client" or "server". Specifies if certificates are
+            flagged for server or client authentication use. See RFC 5280 Section 4.2.1.12 for
+            information about the Extended Key Usage field.
+        :param common_name: The requested CN for the certificate.
+        :param sans: Subject Alternative Name
+        :return: None
+        """
+        if not sans:
+            sans = []
+        logger.info("Received request to create certificate")
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            logger.info(
```

Comment:
I think it's ok for it to be an info. One could deploy the requirer side, have the certificate request in the "on_install" hook, and then relate its charm to a provider. If we change it to `error` this would have trigerred an error in the logs.

Reply:
Yes, that was the idea - error in the logs draws attention. That's also why, in the comment below, I proposed to raise some exception here - if charm implementer want's to request a cert in _on_install_, he could leverage the exception to set charm's state to blocked, until relation is created.